src/v/container/include/container/chunked_vector.h

```diff
@@ -0,0 +1,296 @@
+    using this_type = chunked_vector<T>;
+
+public:
+    using value_type = T;
+    using reference = T&;
+    using const_reference = const T&;
+    using size_type = size_t;
+
+    chunked_vector() noexcept = default;
+    chunked_vector& operator=(const chunked_vector&) noexcept = delete;
+    chunked_vector(chunked_vector&& other) noexcept {
+        *this = std::move(other);
+    }
+    chunked_vector& operator=(chunked_vector&& other) noexcept {
+        if (this != &other) {
+            this->_size = other._size;
+            this->_chunks = std::move(other._chunks);
+            // Move compatibility with std::vector that post move
+            // the vector is empty().
+            other._size = 0;
+        }
+        return *this;
+    }
+    ~chunked_vector() noexcept = default;
+
+    template<typename Iter>
+    requires std::input_iterator<Iter>
+    chunked_vector(Iter begin, Iter end)
+      : chunked_vector() {
+        // Improvement: Write a more efficient implementation for
+        // random_access_iterators
+        for (auto it = begin; it != end; ++it) {
+            push_back(*it);
+        }
+    }
+
+    chunked_vector copy() const noexcept { return *this; }
+
+    void swap(chunked_vector& other) noexcept {
+        std::swap(_size, other._size);
+        std::swap(_chunks, other._chunks);
+    }
+
+    template<class E = T>
+    void push_back(E&& elem) {
+        maybe_add_capacity();
+        _chunks.back().push_back(std::forward<E>(elem));
+        ++_size;
+    }
+
+    template<class... Args>
+    T& emplace_back(Args&&... args) {
+        maybe_add_capacity();
+        _chunks.back().emplace_back(std::forward<Args>(args)...);
+        ++_size;
+        return _chunks.back().back();
+    }
+
+    void pop_back() {
+        vassert(_size > 0, "Cannot pop from empty container");
+        _chunks.back().pop_back();
+        --_size;
+        // Invariant: always keep at least one chunk
+        if (_chunks.back().empty() && _chunks.size() > 1) {
+            _chunks.pop_back();
+        }
+    }
+
+    /*
+     * Replacement for `erase(some_it, end())` but more efficient than n
+     * `pop_back()s`
+     */
+    void pop_back_n(size_t n) {
+        vassert(
+          _size >= n, "Cannot pop more than size() elements in container");
+
+        if (_size == n) {
+            clear();
+            return;
+        }
+
+        _size -= n;
+
+        // Invariant: always keep at least one chunk
+        while (n >= _chunks.back().size() && _chunks.size() > 1) {
+            n -= _chunks.back().size();
+            _chunks.pop_back();
+        }
+
+        for (size_t i = 0; i < n; ++i) {
+            _chunks.back().pop_back();
+        }
+    }
+
+    const T& operator[](size_t index) const {
+        vassert(index < _size, "Index out of range {}/{}", index, _size);
+        auto& frag = _chunks.at(index / elements_per_chunk);
+        return frag.at(index % elements_per_chunk);
```

Comment:
`.at()` vs `operator[size_t]` is probably worth explaining.

Reply:
Thanks yes, I'm a little surprised we do bounds checking here, since std::vector would not by default.

Reply:
Removed this.